src/term/transform/resolve_ctrs_in_pats.rs

```diff
@@ -20,59 +20,67 @@ impl Book {
 
 impl Pattern {
   pub fn resolve_ctrs(&mut self, is_ctr: &impl Fn(&Name) -> bool) {
-    match self {
-      Pattern::Var(Some(nam)) => {
-        if is_ctr(nam) {
-          *self = Pattern::Ctr(nam.clone(), vec![]);
+    let mut to_resolve = vec![self];
+
+    while let Some(pat) = to_resolve.pop() {
+      match pat {
+        Pattern::Var(Some(nam)) => {
+          if is_ctr(nam) {
+            *pat = Pattern::Ctr(nam.clone(), vec![]);
+          }
         }
-      }
-      Pattern::Ctr(_, args) | Pattern::Lst(args) => {
-        for arg in args {
-          arg.resolve_ctrs(is_ctr);
+        Pattern::Ctr(_, args) | Pattern::Lst(args) => {
+          for arg in args {
+            to_resolve.push(arg);
+          }
+        }
+        Pattern::Var(None) => (),
+        Pattern::Num(_) => (),
+        Pattern::Tup(fst, snd) => {
+          to_resolve.push(fst);
+          to_resolve.push(snd);
         }
-      }
-      Pattern::Var(None) => (),
-      Pattern::Num(_) => (),
-      Pattern::Tup(fst, snd) => {
-        fst.resolve_ctrs(is_ctr);
-        snd.resolve_ctrs(is_ctr);
       }
     }
   }
 }
 
 impl Term {
   pub fn resolve_ctrs_in_pats(&mut self, is_ctr: &impl Fn(&Name) -> bool) {
-    match self {
-      Term::Let { pat, val, nxt } => {
-        pat.resolve_ctrs(is_ctr);
-        val.resolve_ctrs_in_pats(is_ctr);
-        nxt.resolve_ctrs_in_pats(is_ctr);
-      }
-      Term::Mat { matched: scrutinee, arms } => {
-        scrutinee.resolve_ctrs_in_pats(is_ctr);
-        for (pat, body) in arms {
+    let mut to_resolve = vec![self];
+
+    while let Some(pat) = to_resolve.pop() {
+      match pat {
+        Term::Let { pat, val, nxt } => {
           pat.resolve_ctrs(is_ctr);
-          body.resolve_ctrs_in_pats(is_ctr);
+          to_resolve.push(val);
+          to_resolve.push(nxt);
         }
+        Term::Mat { matched: scrutinee, arms } => {
```

Comment:
Can you rename the variable to `matched`? I must have missed some places when I renamed the field